File: 22. Generate Parentheses: O(2^(2n)).py
# 22. Generate Parentheses
# https://leetcode.com/problems/generate-parentheses/description/

# O(2^(2n) / sqrt(n)) algorithm for generating well-formed parentheses using backtracking:
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        # Initialize an empty stack and result' list:
        stack = []
        valid_parentheses = []

        # Do backtracking DFS while abandoning non-well-formed parentheses:
        def backtrack(open_count, closed_count):
            # Save if number of "(" and ")" match pair count (and are in sequence):
            if open_count == closed_count == n:
                valid_parentheses.append("".join(stack))
            
            # Only add "(" to string if pair count not exhausted:
            if open_count < n:
                stack.append("(")
                backtrack(open_count + 1, closed_count)
                stack.pop()
            
            # Only add ")" to string if there are "(" to close properly:
            if closed_count < open_count:
                stack.append(")")
                backtrack(open_count, closed_count + 1)
                stack.pop()
        
        # Recursively DFS from empty string until well-formed parentheses length 2n:
        backtrack(0, 0)

        # Return list of well-formed parentheses with n number of "(" and ")":
        return valid_parentheses

File: 22. Generate Parentheses: O(2^(2n)).rb
# 22. Generate Parentheses
# https://leetcode.com/problems/generate-parentheses/description/

# @param {String} s
# @return {Boolean}

# O(n) algorithm for checking balanced parentheses using stack (from 20. Valid Parentheses):
def valid_parentheses(s)
    # Initialize an empty stack:
    stack = []
    
    # Iterate through all brackets in string::
    s.each_char do |char|
        # When opening bracket is found, push it into stack:
        if ['('].include? char
            stack.push char
        # When closing bracket is found, compare it to opening bracket from stack:
        else
            # If there is no opening bracket to match, return false:
            return false if stack.empty?
        
            # If opening and closing brackets mismatch, return false:
            top = stack.pop
            return false if char == ')' && top != '('
        end
    end
    
    # If brackets didn't mismatch and no unclosed brackets left (stack is empty), return true:
    stack.empty?
end

# @param {Integer} n
# @return {String[]}

# Brute force O(2^(2n)) algorithm for generating well-formed parentheses from all strings:
def generate_parenthesis(n)
    # Initialize queue that will hold all possible strings length 1 to 2n:
    queue = ['']
    valid_strings = []

    while !queue.empty?
        # Dequeue an item from queue (dequeues all strings after length greater than 2n):
        s = queue.shift
        
        # If the length of s is 2n and is valid parenthesis, add to result:
        if s.length == 2 * n && valid_parentheses(s)
            valid_strings.append(s)
        # If the length of s is less than 2n, append '(' and ')' to s and enqueue:
        elsif s.length < 2 * n
            queue.push(s + '(')
            queue.push(s + ')')
        end
    end

    return valid_strings
end
